### PR TITLE
[Hackathon No.18] 新增frexp api

### PR DESCRIPTION
**Pull request by Zheng-Bicheng**
### PR types
New features

### PR changes
APIs

### Describe
新增frexp api
提案：https://github.com/PaddlePaddle/community/pull/214
中文文档：https://github.com/PaddlePaddle/docs/pull/5276

### CODE REVIEW

**Comment by Zheng-Bicheng** [comment]
[提案链接](https://github.com/PaddlePaddle/community/pull/214)

**Review by tizhou86 — COMMENTED** [review]
请在python/paddle/__init__.py中也添加api声明，可以参考这个：https://github.com/PaddlePaddle/Paddle/pull/44853/files#diff-dd4ce82a4831b8c3f62aa1717373fce4dd62e1f42540bc8395df5e3c447a0475 ，然后再过一下ci

**Comment by Zheng-Bicheng** [comment]
> 请在python/paddle/**init**.py中也添加api声明，可以参考这个：https://github.com/PaddlePaddle/Paddle/pull/44853/files#diff-dd4ce82a4831b8c3f62aa1717373fce4dd62e1f42540bc8395df5e3c447a0475 ，然后再过一下ci

已经按照您的要求进行了修改，您再看一下

**Comment by Zheng-Bicheng** [comment]
> 请在python/paddle/**init**.py中也添加api声明，可以参考这个：https://github.com/PaddlePaddle/Paddle/pull/44853/files#diff-dd4ce82a4831b8c3f62aa1717373fce4dd62e1f42540bc8395df5e3c447a0475 ，然后再过一下ci

还需要做什么改进嘛

**Comment by Zheng-Bicheng** [comment]
> 请在python/paddle/**init**.py中也添加api声明，可以参考这个：https://github.com/PaddlePaddle/Paddle/pull/44853/files#diff-dd4ce82a4831b8c3f62aa1717373fce4dd62e1f42540bc8395df5e3c447a0475 ，然后再过一下ci

有时间能帮忙再看一下吗

File: python/paddle/fluid/tests/unittests/test_frexp_api.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#
+#     mantissa = paddle.where((x < 0), mantissa * -1, mantissa)
+#     return mantissa, exponent
+
+
+class TestFrexpAPI(unittest.TestCase):
+
+    def setUp(self):
+        np.random.seed(1024)
+        self.x_np = np.random.uniform(-3, 3, [10, 12]).astype('float32')
```
Comment (tizhou86) [inline python/paddle/fluid/tests/unittests/test_frexp_api.py]: 需要添加float64测试案例，可以参考一下：https://github.com/PaddlePaddle/Paddle/pull/44853/files 的写法

File: python/paddle/fluid/tests/unittests/test_frexp_api.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        self.place = paddle.CUDAPlace(0) if paddle.is_compiled_with_cuda() \
+            else paddle.CPUPlace()
+
+    # 静态图单测
+    def test_static_api(self):
+        # 开启静态图模式
+        paddle.enable_static()
+        with paddle.static.program_guard(paddle.static.Program()):
+            input_data = paddle.fluid.data('X', [10, 12])
```
Comment (tizhou86) [inline python/paddle/fluid/tests/unittests/test_frexp_api.py]: 这儿可以多加一些corner case。

File: python/paddle/fluid/tests/unittests/test_frexp_api.py (hunk 1 above)
Comment (Zheng-Bicheng) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: 已经添加，正在测试

File: python/paddle/fluid/tests/unittests/test_frexp_api.py (hunk 2 above)
Comment (Zheng-Bicheng) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: corner case指的是什么，我参考https://github.com/PaddlePaddle/Paddle/pull/44853/files 加了一个shape不一样的输入数据可以吗？

Comment (tizhou86) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: 恩，可以

File: python/paddle/fluid/tests/unittests/test_frexp_api.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    """
+
+    def set_input(self):
+        self.x_np_1 = np.random.uniform(-3, 3, [10, 12]).astype('float64')
+        self.x_np_2 = np.random.uniform(-1, 1, [4, 5, 2]).astype('float64')
+
+
+if __name__ == "__main__":
+    unittest.main()
+    # a = TestFrexpAPI()
```
Comment (tizhou86) [inline python/paddle/fluid/tests/unittests/test_frexp_api.py]: 这儿注释的部分可以删除

Comment (Zheng-Bicheng) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: 已经删除无用的注释以及测试用的代码，只保留了测试代码，正在测试中

Comment (Zheng-Bicheng) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: CI都过了，您在看看有没有什么问题，没问题能给个锁定冠军吗，我就去写文档了😬

Comment (Zheng-Bicheng) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: 写文档的pr也需要关联提案嘛？

Comment (Ligoml) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: 文档链接贴在这个pr的描述中就可以，不需要关联提案~

Comment (Zheng-Bicheng) [reply python/paddle/fluid/tests/unittests/test_frexp_api.py]: > 文档链接贴在这个pr的描述中就可以，不需要关联提案~

那这边的代码还有什么问题吗？

**Comment by tizhou86** [comment]
LGTM

**Comment by Zheng-Bicheng** [comment]
https://github.com/PaddlePaddle/docs/pull/5276

**Comment by Zheng-Bicheng** [comment]
> LGTM

两份文档都补充完成了，您看一下

File: python/paddle/tensor/math.py [hunk 4]
```diff
@@ -4819,6 +4819,45 @@ def sgn(x, name=None):
...
 
+
+def frexp(x):
+    """
+        The function used to decompose a floating point number into mantissa and exponent.
+
+        Args:
+            x (Tensor): The input tensor, which data type should be float32, float64.
+
+        Returns:
```
Comment (Ligoml) [inline python/paddle/tensor/math.py:4887]: 返回值是一个list还是两个Tensor呢？看示例代码是list，但是这里的描述是两个独立的Tensor？

File: python/paddle/tensor/math.py [hunk 5]
```diff
@@ -4819,6 +4819,45 @@ def sgn(x, name=None):
...
+            mantissa: A mantissa Tensor for real input, shape and data type are same as input.
+            exponent: A mantissa Tensor for real input, shape and data type are same as input.
+
+        Examples:
+            .. code-block:: Python
+
+                import paddle
+
+                x = paddle.to_tensor([[1,2,3,4]])
+                print(paddle.sgn(x))
```
Comment (Ligoml) [inline python/paddle/tensor/math.py]: sgn？

File: python/paddle/tensor/math.py [hunk 6]
```diff
@@ -4819,6 +4819,45 @@ def sgn(x, name=None):
     else:
         return paddle.sign(x)
 
+
+def frexp(x):
```
Comment (Ligoml) [inline python/paddle/tensor/math.py:4880]: 参数是否需要增加name呢？@tizhou86

File: python/paddle/tensor/math.py (hunk 4 above)
Comment (Zheng-Bicheng) [reply python/paddle/tensor/math.py:4887]: 已经修复，应该是一个元组，下意识写成列表了

File: python/paddle/tensor/math.py (hunk 5 above)
Comment (Zheng-Bicheng) [reply python/paddle/tensor/math.py]: 已经修复
